torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -531,45 +528,28 @@ class CppMicroGemmAMX(CppMicroGemm):
     alignas(4096) {{input_t}} dequantized_B_buf[buf_size];
     {%- endif %}
 
-    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
-
-    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
-        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
-        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
-        b_bf16.store(dst);
-    };
-
-    auto load_B_tile = [&]({{input2_t}}* B_ptr, int idx, int num_b_rows) {
-        {{input_t}}* base_addr = dequantized_B_buf + idx;
-        {{kernel.unroll_pragma(8)}}
-        for (int i = 0; i < num_b_rows; i++) {
-            load_B_row(
-                B_ptr + i * b_tile_ptr_stride,
-                base_addr + i * 32
-            );
-        }
-    };
-    auto load_dequantized_B = [&](int n) {
+    auto load_dequantized_B = [&](int tile_idx) {
         // Load a tile of B & cache it in L1D.
-        {{kernel.unroll_pragma(4)}}
-        for (int k = 0; k < K; k += {{block_k}}) {
-            int num_b_rows = (k < last_k_offset) ? 16 : tail_k_size;
-            {{kernel.unroll_pragma(2)}}
-            for (int tile_col = 0; tile_col <= 1; tile_col++) {
-                load_B_tile(
-                    const_cast<{{input2_t}}*>(B) + n + k * ldb + tile_col * {{16 * vnni_size}},
-                    (k / {{block_k // 2}} + tile_col) * num_elements_per_b_tile,
-                    num_b_rows
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n,
+        // then B would be some consecutive tiles sized [k, block_n] elements each, so the
+        // first element of a subsequent [K, block_n] tile would be right after the last
+        // element of the previous tile sized [K x block_n] elements.
```

Comment:
Hi, `ldb` is not required here (at least for `N == Nr`), since with this implementation, the buffer has int8 -> bf16 data from `B` in the same order as `B`.

In case of `N != Nr`, though, I guess all the `K * N` elements would be contiguous in `B`, but please confirm. Thanks!

Reply:
The API semantics is that `B` is of the shape `[K, N]` with stride `[ldb, 1]` where `ldb` is not necessarily equal to `N`. I don't think it is correct to assume `B` is contiguous.

Reply:
> I don't think it is correct to assume B is contiguous.

Thanks! Please clarify if you mean `B` tiles may not be contiguous if `N != block_n` (which is currently not the case).

 `B` is currently contiguous due to `N` in the micro-kernel being equal to `block_n`, and weight pre-packing of the whole B matrix results in `B` tiles (used by the micro-kernel) being contiguous.

> `ldb` is not necessarily equal to `N`

Does this mean that to support `N != Nr` in the micro-kernel, weight-packing implementation would not have to be changed? Thanks!

Reply:
The implementation (& the comment) does seem incorrect for `N != block_n`, because if a B-tile is shaped `[K, N]`, then `N != block_n` can't be supported by multiple `[K, block_n]` blocks being contiguous (at the granularity of blocks, although all the elements of the blocks would also be contiguous) in memory because that case would actually be like `[K * constant, block_n]`.

Please clarify if this is close to what you have in mind for `N != block_n` (I'm guessing it'd also require a change in the weight-packing implementation):

<img width="723" alt="image" src="https://github.com/user-attachments/assets/36f3938f-8c70-44d4-a806-f40d8d45b2f9">

Reply:
@leslie-fang-intel, I feel like `N != block_n` would make dequantizing & caching a `[K, block_n]` block hard if `N != block_n`. That's because the each row of a `K, block_n` block would be discontiguous (the next row would not be adjacent to it).

The simplified dequantization code from your patch won't work in that case (`block_n != 32` and `block_n != N`).
So, it looks like we'd have to revert to the previous implementation (but adding another patch for #140208).

Reply:
Yes, this could be a possible scenario. Thanks.

Reply:
Thanks, I revised for such a case

Reply:
@leslie-fang-intel, looks like the approach to dequantize 32 elements at a time fails when `N != Nr` and `block_n` is either 16 or 48.

That's because we can't dequantize 32 elements at a time in that case, since they're not contiguous. 
We can only dequantize 16 at a time in case of `block_n` being 16, 
and 32 followed by 16 (or vice-versa) in case of `block_n` being 48